programs/drift/src/math/liquidation.rs

```diff
@@ -295,3 +300,147 @@ pub fn calculate_cumulative_deposit_interest_delta_to_resolve_bankruptcy(
         .safe_div(total_deposits)
         .or(Ok(0))
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DeleverageUserStats {
+    pub base_asset_amount: i64,
+    pub quote_asset_amount: i64,
+    pub quote_break_even_amount: i64,
+    pub free_collateral: i128,
+}
+
+pub fn calculate_perp_market_deleverage_payment(
+    loss_to_socialize: i128,
+    deleverage_user_stats: DeleverageUserStats,
+    market: &PerpMarket,
+    oracle_price: i64,
+) -> DriftResult<i128> {
+    let mean_short_entry_basis = if market.amm.base_asset_amount_short != 0 {
+        market
+            .amm
+            .quote_break_even_amount_short
+            .safe_mul(PRICE_TIMES_AMM_TO_QUOTE_PRECISION_RATIO_I128)?
+            .safe_div(market.amm.base_asset_amount_short)?
+    } else {
+        0
+    };
+
+    let mean_short_pnl_per_base = mean_short_entry_basis
+        .safe_sub(oracle_price.cast()?)?
+        .safe_div(PRICE_TO_QUOTE_PRECISION_RATIO.cast()?)?;
+
+    let mean_long_entry_basis = if market.amm.base_asset_amount_long != 0 {
+        market
+            .amm
+            .quote_break_even_amount_long
+            .safe_mul(PRICE_TIMES_AMM_TO_QUOTE_PRECISION_RATIO_I128)?
+            .safe_div(market.amm.base_asset_amount_long)?
+    } else {
+        0
+    };
+
+    let mean_long_pnl_per_base = mean_long_entry_basis
+        .safe_add(oracle_price.cast()?)?
+        .safe_div(PRICE_TO_QUOTE_PRECISION_RATIO.cast()?)?;
+
+    let mean_entry_basis: i128 = if deleverage_user_stats.base_asset_amount > 0 {
+        mean_long_entry_basis
+    } else if deleverage_user_stats.base_asset_amount < 0 {
+        mean_short_entry_basis
+    } else {
+        0
+    };
+
+    let user_entry_basis: i128 = if deleverage_user_stats.base_asset_amount != 0 {
+        PRICE_TIMES_AMM_TO_QUOTE_PRECISION_RATIO_I128
+            .safe_mul(deleverage_user_stats.quote_break_even_amount.cast()?)?
+            .safe_div(deleverage_user_stats.base_asset_amount.cast()?)?
+    } else {
+        0
+    };
+
+    let exit_basis = oracle_price
+        .safe_mul(deleverage_user_stats.base_asset_amount.signum().cast()?)?
+        .safe_div(PRICE_TO_QUOTE_PRECISION_RATIO.cast()?)?;
+
+    let mean_pnl_per_base = mean_entry_basis.safe_add(exit_basis.cast()?)?;
+    let user_pnl_per_base = user_entry_basis.safe_add(exit_basis.cast()?)?;
+
+    let profit_above_mean = if deleverage_user_stats.base_asset_amount == 0 {
+        msg!("user has no base");
+        if deleverage_user_stats.quote_asset_amount != 0 {
+            let mean_quote = (market.amm.base_asset_amount_short)
+                .safe_add(market.amm.base_asset_amount_long)?
+                .safe_mul(oracle_price.cast()?)?
+                .safe_div(PRICE_TIMES_AMM_TO_QUOTE_PRECISION_RATIO_I128)?
+                .safe_add(calculate_net_user_cost_basis(&market.amm)?)?
+                .safe_add(loss_to_socialize)?
+                .safe_div(market.number_of_users.max(1).cast()?)?;
```

Comment:
if number_of_users == 0 and the user still has quote, we're in a corrupted state, right?

Reply:
yep this was for the old number_of_users logic, will remove max(1)